std/file.d

```diff
@@ -450,22 +459,112 @@ version (Windows) private void[] readImpl(scope const(char)[] name, scope const(
             HANDLE.init);
     auto h = trustedCreateFileW(namez, defaults);
-
     cenforce(h != INVALID_HANDLE_VALUE, name, namez);
     scope(exit) cenforce(trustedCloseHandle(h), name, namez);
-    ulong fileSize = void;
-    cenforce(trustedGetFileSize(h, fileSize), name, namez);
-    size_t size = min(upTo, fileSize);
-    auto buf = () @trusted { return GC.malloc(size, GC.BlkAttr.NO_SCAN)[0 .. size]; } ();
+
+              ulong  fileSize     = void;
+    immutable bool   fileSizeOkay = trustedGetFileSize(h, fileSize);
+    immutable size_t maxSize      = fileSizeOkay ? min(fileSize, upTo) : upTo;
+
+    //writef("readFile %s: fileSize %u.\n", name, fileSize);
+
+    // Shamelessly reproducing the rest of the function from Posix version, with adaptations kept to a minimum:
+
+    enum size_t
+        minInitialAlloc       = 0x1000,
+        maxInitialAlloc       = size_t.max / 2,
+        sizeIncrement         = 0x1000,
+        maxSlackMemoryAllowed = 0x1000,
+        unknownPurposeSize    = 0; // Changed from 1 in Posix version, to support callers requesting 4 KB etc.
+
+    immutable size_t initialAlloc = min
+    (
+        upTo,
+        to!size_t
+        (
+            fileSizeOkay ? min(fileSize + unknownPurposeSize, maxInitialAlloc) : minInitialAlloc
+        )
+    );
+
+    enum attr = GC.BlkAttr.NO_SCAN;
+    auto buf = () @trusted { return GC.malloc(initialAlloc, attr)[0 .. initialAlloc]; } ();
+    //writef ("buf %16Xh + %8Xh = %16Xh.\n", buf.ptr, buf.length, buf.ptr + buf.length);
 
     scope(failure)
     {
         () @trusted { GC.free(buf.ptr); } ();
     }
 
-    if (size)
-        cenforce(trustedReadFile(h, &buf[0], size), name, namez);
-    return buf[0 .. size];
+    auto size = checked(size_t(0));
+    for (;;)
+    {
+        immutable size_t wanted     = (min(buf.length, maxSize) - size).get;
+                  ulong  actual     = void;
+        immutable bool   actualOkay = trustedReadFile(h, buf.ptr + size.get, wanted, actual);
+        cenforce(actualOkay, name, namez);
+
+        if (! actual)
+            break;
+
+        size += actual;
+
+        assert(size <= maxSize);
+        if (size >= maxSize)
+            break;
+
+        assert(size <= buf.length);
+        if (size == buf.length)
+        {
+            immutable newAlloc = size + sizeIncrement;
+            buf = GC.realloc(buf.ptr, newAlloc.get, attr)[0 .. newAlloc.get];
+            //writef ("buf %16Xh + %8Xh = %16Xh.\n", buf.ptr, buf.length, buf.ptr + buf.length);
+        }
+    }
+
+    if (buf.length - size >= maxSlackMemoryAllowed)
+    {
+        buf = GC.realloc(buf.ptr, size.get, attr)[0 .. size.get];
+        //writef ("buf %16Xh + %8Xh = %16Xh.\n", buf.ptr, buf.length, buf.ptr + buf.length);
+    }
+
+    return buf[0 .. size.get];
+}
+
+@safe unittest
+{
+    import std.stdio;
+
+    enum size_t numIdenticalChars = 3989; // Dummy uncomfortable value (not a multiple of any beautiful number).
```

Comment:
So... a prime number?

Reply:
Indeed, the largest prime number less than 4096.

The various magical numbers which appear in the new code might not be the best possible choices. My (rather vague) plan has been to avoid working with buffers of 4096, both in the "write" code and in the "read" code. It felt like too round a number to use everywhere.

By the way:
Would it make sense to allow the caller of "std.file.read" to optionally specify parameters (e.g. chunk size, maximum slack size etc.) -- e.g. in an optional "struct" ?

Reply:
>  Would it make sense to allow the caller of "std.file.read" to optionally specify parameters (e.g. chunk size, maximum slack size etc.) -- e.g. in an optional "struct" ?

I don't think so. Reading from a non-seekable file is by itself uncommon, and the requirement to be able to tweak such settings would be even more so. For uncommon use cases, a custom implementation satisfying the particular needs is usually more appropriate. It's also rather simple to implement using e.g. `File.byChunk`.